### DISCUSSION TITLE
Rethrow undeclared exceptions

### DISCUSSION THREAD

**Discussion by ttulka** [spring-projects/spring-retry#263] [Q&A] [answered]
Not sure if this is a "works as designed" problem, but it seems to be useful to support this use-case. 
So far, I was not able to make it work:

I need the retry mechanism to be triggered only for declared exceptions and let others be rethrown.

With this code:

```java
@Retryable(include = {ToRetryException.class}, recover = "recoverRetrieveFromExternal")
public String retrieveFromExternal(String arg) {
    return externalService.retrieve(arg);
}
@Recover
private String recoverRetrieveFromExternal(Throwable t, String arg) {
    return "Recover: " + arg;
}
```

When an exception different from `ToRetryException` is thrown by `externalService.retrieve(arg)` the recover fallback is used.

With `stateful = true`, all exceptions including `ToRetryException` are rethrown.

Basically, I need this test suite to pass:

```java
@Test
void ok() {
    when(externalService.retrieve(anyString())).thenReturn("TEST");
    assertThat(client.retrieveFromExternal("")).isEqualTo("TEST");
}
@Test
void rethrows_the_exception() {
    when(externalService.retrieve(anyString())).thenThrow(new RuntimeException());
    assertThrows(RuntimeException.class, () -> client.retrieveFromExternal(""));
}
@Test
void retries_and_recovers() {
    when(externalService.retrieve(anyString())).thenThrow(new ToRetryException());
    assertThat(client.retrieveFromExternal("test")).isEqualTo("Recover: test");
}
```

Is there a way to do that? Thanks!

**Comment by garyrussell** [comment, marked as answer]
Don't use GitHub issues to ask questions; they are for bug reports or asking for new features; ask questions on Stack Overflow or use the Discussions tab.

Change the recover method so that it only handles `ToRetryException`, not all `Throwable`s.

**Reply by ttulka** [reply]
@garyrussell  Many thanks!

Changing the recover method is problematic because 
1. `ExhaustedRetryException` is thrown instead of `ToRetryException`
2.  Multiple exception types (my real-world usecase) in `include` need multiple recover methods which would be tedious

Would a new attribute like `boolean rethrow` be a thinkable enhancement? 

```java
@Retryable(include = {FooException.class, BarException.class}, rethrow = true)
// rethrows immediately any exception not included in the declaration
```

**Reply by garyrussell** [reply]
The `RetryTemplate` already has the property `throwLastExceptionOnExhausted` (false by default); setting it via the annotation is not currently supported (we could do that).

The actual exception is in the `cause` of the `ExhaustedRetryException` so you could use that in the caller (not optimal, I agree).

Another work around would be to wire up your own `RetryOperationsInterceptor` with a suitably configured `RetryTemplate` (set it via the `interceptor` property in `@Retryable`).

**Reply by garyrussell** [reply]
However, in this case, the `ExhaustedRetryException` is not thrown by the template, it's thrown by the `RecoverAnnotationRecoveryHandler`, so that would need a tweak too (not impacted by the interceptor work around though).

**Reply by garyrussell** [reply]
https://github.com/spring-projects/spring-retry/issues/264

**Reply by ttulka** [reply]
Yeah, I had started writing my own interceptor originally. Unfortunately, it means losing a lot of elegance in comparison to `@EnableRetry` (and `RetryConfiguration` with `AnnotationAwareRetryOperationsInterceptor`). Some flexible solution provided by the API would be great.

I am not sure about `throwLastExceptionOnExhausted` as it seems to take action only for a *stateful* interceptor: `RetryTemplate.java:539`
Using a stateful interceptor is not wanted in my use-case.

I need something like the following:

```java
protected <T> T handleRetryExhausted(
    RecoveryCallback<T> recoveryCallback, RetryContext context, RetryState state, RetryPolicy retryPolicy)
        throws Throwable {
    context.setAttribute(RetryContext.EXHAUSTED, true);
    if (state != null && !context.hasAttribute(GLOBAL_STATE)) {
        this.retryContextCache.remove(state.getKey());
    }
    if (this.rethrow && !retryPolicy.retryForException(context.getLastThrowable())) {
        throw context.getLastThrowable();
    }
    if (recoveryCallback != null) {
        T recovered = recoveryCallback.recover(context);
        context.setAttribute(RetryContext.RECOVERED, true);
        return recovered;
    }
    if (state != null) {
        this.logger.debug("Retry exhausted after last attempt with no recovery path.");
        rethrow(context, "Retry exhausted after last attempt with no recovery path");
    }
    throw wrapIfNecessary(context.getLastThrowable());
}
```

Where `rethrow` is the new flag propagated from `@Retryable` and `retryPolicy.retryForException` returns `true` when the exception is registered as retryable (by `include` in `@Retryable`).

Does it make sense? If so, I can try to create a PR.

**Reply by garyrussell** [reply]
It generally looks ok but, unfortunately, this project is still based on Java6 so we can't add a new method to the `RetryPolicy` (can't use `default` in Java 6) because it is a breaking change - we will probably be moving to Java 17 in the 2.0 line. It would need to be a new interface (with an instanceof check).

Also, although not needed for your use case, for consistency, we should add a flag similar to `throwLastExceptionOnExhausted` to the `RecoverAnnotationRecoveryHandler` (e.g. `throwLastExceptionWhenNoRecoverMethod`) so it throws the raw exception instead of wrapping it in a `ExhaustedRetryException` when there is no matching `@Recover` method.

**Reply by ttulka** [reply]
I see. In that case I guess `!((SimpleRetryPolicy) retryPolicy).retryForException(context.getLastThrowable())` would do the job for now as `SimpleRetryPolicy` is the default implementation in template anyway.

I have created a PR: https://github.com/spring-projects/spring-retry/pull/265

I will align `RecoverAnnotationRecoveryHandler` afterward.

**Reply by ttulka** [reply]
Hmm, implemented like that, I think no changes in `RecoverAnnotationRecoveryHandler` are necessary as `recoveryCallback.recover(context)` is not called at all when the rethrow flag is set.

**Reply by garyrussell** [reply]
Right; but what about a retryable exception that has no compatible `@Recover` method? - we need consistency - when rethrow is true it should rethrow the raw exception and not wrap it, for consistency with not-retryable exceptions.

**Reply by ttulka** [reply]
Implemented in the second commit of https://github.com/spring-projects/spring-retry/pull/265

**Comment by garyrussell** [comment]
Converting to a discussion.

**Reply by ttulka** [reply]
Sorry about that!
